File: recursion/bits.cpp
#include <iostream>
#include <vector>
using namespace std;

int offToOn(int num, int k) {
    int mask = (1 << k);
    return (num | mask);
}
int onToOff01(int num, int k) {
    int mask = (~(1 << k));     //1s complement
    return (num & mask);
}

int onToOff02(int num, int k) {
    int mask = (1 << k);     
    if((num & mask) == 0) return num;
    else return (num ^ mask);
}

int countAllSetBits_02(unsigned num) {
    int count = 0;
    while(num != 0) {
        if((num & 1) == 1) count++;
        num = (num >> 1);
    }
    return count;
}



int main() {
    // cout << offToOn(10, 0);
    // cout << onToOff01(5, 2);
    // cout << onToOff02(5, 2);
    // cout << (onToOff01(5, 2) == onToOff02(5, 2));
    // cout << countAllSetBits_02(121223);
    return 0;
}

File: recursion/bits.java
public class bits {
    public static void main(String[] args) {
        // System.out.println(countAllSetBits_03(-1));
        int[] arr = {2, 2, 3, 4, 2, 2, 4};
        System.out.println(uniqueNumbInArray(arr));
    }

    public static int countAllSetBits(int num) {
        int count = 0;
        for(int i = 0; i < 32; i++) {
            int mask = (1 << i);
            if((num & mask) != 0)   count++;
        }
        return count;
    }

    public static int countAllSetBits_02(int num) {
        int count = 0;
        while(num != 0) {
            if((num & 1) == 1) count++;
            num = (num >>> 1);
        }
        return count;
    }

    public static int countAllSetBits_03(int num) {
        int count = 0;
        while(num != 0) {
            count++;
            num &= (num - 1);
        }
        return count;
    }

    public static int uniqueNumbInArray(int[] arr) {
        int res = 0;
        for(int ele : arr) {
            res = (res ^ ele);
        }
        return res;
    }
}